tests/unit/lib/services.spec.js

```diff
@@ -0,0 +1,94 @@
+const assert = require('assert');
+const sinon = require('sinon');
+const { versioning } = require('arsenal');
+
+const services = require('../../../lib/services');
+const { DummyRequestLogger } = require('../helpers');
+
+const { VersionId } = versioning.VersioningConstants;
+
+describe('services', () => {
+    const log = new DummyRequestLogger();
+    const bucketName = 'test-bucket';
+    const objectKey = 'test-object';
+
+    afterEach(() => {
+        sinon.restore();
+    });
+
+    describe('findObjectVersionByUploadId', () => {
+        let getObjectListingStub;
+
+        beforeEach(() => {
+            getObjectListingStub = sinon.stub(services, 'getObjectListing');
+        });
+
+        it('should call getObjectListing with an optimized prefix', done => {
+            getObjectListingStub.yields(null, { Versions: [], IsTruncated: false });
+
+            services.findObjectVersionByUploadId(bucketName, objectKey, 'any-upload-id', log, err => {
+                assert.ifError(err);
+                sinon.assert.calledOnce(getObjectListingStub);
+
+                const listParams = getObjectListingStub.getCall(0).args[1];
+                const expectedPrefix = `${objectKey}${VersionId.Separator}`;
+                assert.strictEqual(listParams.prefix, expectedPrefix);
+                assert.strictEqual(listParams.listingType, 'DelimiterVersions');
+                done();
+            });
+        });
+
+        it('should correctly handle pagination', done => {
```

Comment:
this test is more "should read all pages", but it does not fully check we correctly handle pagination...
e.g. we have have other tests to check that we don't call later pages if we find the entry we are looking for, or that we are able indeed to find an object on a later page...

(I can imagine 5 cases : not found, fond on only page, found on first page of many, found on last page of many, found on "middle" page of many... which could be reduced based on the conditions in the code, but there are a bit too many nested variables/lambdas for me to be completely sure)

Reply:
I added these tests.